django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,130 @@
+from django.db.models import Q
+
+
+def user_can_edit_chants_in_source(user, source) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_is_proofreader(user) -> bool:
+    """
+    Checks if the user is a proofreader.
+    Used in UserSourceListView.
+    """
+    return user.groups.filter(Q(name="project manager") | Q(name="editor")).exists()
+
+
+def user_can_proofread_chants_in_source(user, source_id) -> bool:
+    """
+    Checks if the user can access the proofreading page of a given Source.
+    Used in ChantProofreadView.
+    """
+    if user.is_anonymous:
+        return False
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_id_editor: bool = user.groups.filter(name="editor").exists()
+
+    if (user_is_project_manager) or (user_id_editor and user_is_assigned_to_source):
+        return True
+    return False
+
+
+def user_can_view_unpublished_source(user, source) -> bool:
+    """
+    Checks if the user can view unpublished Sources or Chants/Sequences belonging
+    to unpublished sources on the site.
+    Used in ChantDetail, SequenceDetail, and SourceDetail views.
+    """
+    user_is_authenticated: bool = user.is_authenticated
+    if (
+        (source is not None)
```

Comment:
I feel like this is not the right place for this check - we should check that `source` is actually a source before we pass it in to this function.

Annotations on parameters for all these functions could help. If we had `user_can_view_unpublished_source(user: User, source: Source)`, it would be clear to future developers that they can't pass a `None` as an argument (or, if it's intended that you can pass a `None` in, it should be `source: Optional[Source]`).

Reply:
recommending:
```python
def user_can_view_source(user: User, source: Source) -> bool:
    return source.published or user.is_authenticated

Reply:
I did try this at first locally but there were Sequences in the database that didn't belong to a source which would lead to an error when trying to get to the SequenceDetailView, which is why we need the extra check for Source is not None.

The other option would be to add another function to check if a user can view a sequence belonging to an unpublished source.

Reply:
Huh. I just checked now, and it does seem we can have chants/sequences without sources. Surprising, but I guess this is an OldCantus legacy data thing.

In this case, it sounds like we want a `user_can_view_chant` (and possibly `..._sequence`) function, and handle the case where a chant/sequence doesn't have a source there. If we've got a function called `user_can_view_source`, we should use it to check whether a particular user can view a particular source detail page, and not as a substitute to answer another question like whether a particular user can view a particular sequence detail page.

Right. So I think `views/sequence.py` is the right place to handle Sequences without Sources:

Reply:
Also - we may want to check in with Debra about whether to display chants/sequences without sources. In OC, there was a boolean `visible` property (sources could definitely have it, and possibly chants/sequences could too), which we got rid of, rolling everything into our `published`. Since all our visibility-related logic is based on a chant/sequence's source, we may not want to default to displaying chants to everyone on the grounds that they don't belong to an unpublished source.